src/net/sourceforge/kolmafia/persistence/MallPriceDatabase.java

```diff
@@ -86,12 +87,11 @@ private static int updatePricesFromSource(String filename) {
         if (!ItemDatabase.isTradeable(id)) continue;
         Price p = MallPriceDatabase.prices.get(id);
         if (p == null) {
-          MallPriceDatabase.prices.put(id, new Price(price, timestamp));
+          MallPriceDatabase.prices.put(id, new Price(id, price, timestamp));
```

Comment:
For some reason this bothers me.  I know you want prices sorted by id, but could you get the same behavior by making Price implement Comparable and having the associated compareTo just do the expected sort on id?  Or maybe there is a  different implementation that does not include id in price?

Reply:
Either Price has to know its id in order to updated the encoded field when a new price/timestamp is computed, or the id has to be passed into the Price object on each update. I prefer the former, let me know if you disagree.

Reply:
I have no problem including id as part of price.  What I would like to explore is the data structures that contain Price.  If they have been selected, and use id as a key, for various sorts on id, an alternative might be to make price implement Comparable.  I also wonder whether you can accomplish what is needed by never storing anything but encoded prices and decoding when they are used.  I might also explore storing the encoding in a record so there is just one data structure with price info but the places that use it don't always access all fields.  Make sense?

Reply:
Just noticed encoded data is in price so some of my comments no longer apply.

Reply:
Looking at it, I think you don't add id as a field to price.  Instead you add it as a parameter to update.  I think every place update is called the id is already known.

Reply:
I'm not sure why you would need to make Price implement Comparable - could you explain that? `MallPriceDatabase.prices` serves two purposes: it 1) tracks all prices for when we need to save/load from file, and it 2) tracks cached prices for when we do lookups. Purpose 2 means it has to function as a map from some kind of item identifier to the price object. Purpose 1 implies it should be a sorted data structure, or we should keep track of a sorted view of the data structure. The cleanest approach in my view is to make it a sorted map since we're not doing so many lookups that TreeMap vs HashMap matters.

Reply:
OK.  I talked myself out of moving id out of price and then passing id to update.

Reply:
And you win :-)  I tried to implement my own suggestions and decided at best my way was different but not better.